packages/runtime/container-runtime/src/blobAggregationStorage.ts

```diff
@@ -0,0 +1,365 @@
+/*!
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+ import {
+     IDocumentStorageService,
+     ISummaryContext,
+} from "@fluidframework/driver-definitions";
+ import {
+    ISnapshotTree,
+    ISummaryHandle,
+    ISummaryTree,
+    IVersion,
+    SummaryType,
+    ITree,
+} from "@fluidframework/protocol-definitions";
+import {
+    assert,
+    bufferToString,
+    stringToBuffer,
+    unreachableCase,
+    fromUtf8ToBase64,
+    Uint8ArrayToString,
+ } from "@fluidframework/common-utils";
+
+// Gate that when flipped, instructs to compress small blobs.
+// We have to first ship with this gate off, such that we can get to saturation bits
+// that can understand compressed format. And only after that flip it.
+ const gatesAllowPacking = false;
+
+// 2K seems like the sweat spot:
+// The smaller the number, less blobs we aggregate. Most storages are very likely to have notion
+// of minimal "cluster" size, so having small blobs is wasteful
+// At the same time increasing the limit ensure that more blobs with user content are aggregated,
+// reducing possibility for de-duping of same blobs (i.e. .attributes rolled into aggregate blob
+// are not reused across data stores, or even within data store, resulting in duplication of content)
+// Note that duplication of content should not have significant impact for bytes over wire as
+// compression of http payload mostly takes care of it, but it does impact storage size and in-memory sizes.
+const blobCutOffSize = 2048;
+
+/*
+ * Work around for bufferToString having a bug - it can't consume IsoBuffer!
+ * To be removed once bufferToString is fixed!
+*/
+function bufferToString2(blob: ArrayBufferLike, encoding: "utf-8" | "base64"): string {
+    if (blob instanceof Uint8Array) { // IsoBuffer does not have ctor, so it's not in proto chain :(
+        return Uint8ArrayToString(blob, encoding);
+    }
+    return bufferToString(blob, encoding);
+}
+
+ /**
+  * Class responsible for aggregating smaller blobs into one and unpacking it later on.
+  */
+class BlobAggregator {
+    private readonly content: [string, string][]= [];
+
+    public addBlob(key: string, content: string) {
+        this.content.push([key, content]);
```

Comment:
Should we try to dedup contents?  (Small blobs are more likely to be dedup-able too)

Reply:
We could, but note that aggregation (as it's built right now) works on data store level.
So the only real dedupping we can get is 2 attributes for a map DDS when data store has 2 maps ("average" payload has Scriptors that use one directory, 2 maps, and 2 Sequence).

I'd prefer to go for simplicity (given very low gain), but I'll also contradict myself and say that if we believe we need it, we better of building it into unpacking algorithm sooner than later, to have faster deployment timing in future.

Note that bytes over the wire are not affected that much thanks to compression. And in-memory saving is low.
But as a general feature - sure, it sounds great to have it!

I think to punt for now, but open (post-GA) item to do de-duping.